Implement Index and IndexMut for NoiseMap and NoiseImage

SorteKanin opened the pull request:
Implements `Index` and `IndexMut` with 2-dimensional coordinates for NoiseMap as a more ergonomic and efficient alternative to the `get_value` and `set_value` functions, if you are only interested in the data in the map itself. These functions are however preserved for their functionality with the border value. With these trait implementations, one can simply do `map[(x, y)]` to get the value or `map[(x, y)] = new_value` to set the value. This will panic when out of bounds, however.

I've also added `data` and `data_mut` functions to access the underlying vector. This makes it far easier to iterate over the map as you can do `map.data().iter()` or `map.data_mut().iter_mut()`.

Razaekel commented:
Thank you for the contribution!

It looks pretty good, but the Index and IndexMut implementations need bounds checks. Without a bounds check, if somebody had a 10x10 NoiseMap, they could query the position (30,2) and they'd get a value from inside the map back, instead of the border value.

SorteKanin commented:
@Razaekel I see your point - I've adjusted it to perform bounds checks now, but this requires a _tiny_ bit of unsafe code to avoid a double bounds check.

With the way it is now, the get_value and set_value functions don't make a lot of sense to keep (I've kept them in the code, you decide if you want to delete it). I've changed all occurrences to use indexes instead.

This is still a breaking change since IndexMut will panic on out of bounds instead of merely printing an error. There's also the option of returning a mutable reference to the border value but I don't think that makes a lot of sense and that is not what set_value did before either. Do say if there's any more concerns :)

Also, I took the liberty of implementing it for NoiseImage as well.

Razaekel commented:
Please use the safe code version instead, `map.get()` and `map.get_mut()` instead of `get_unchecked()/get_unchecked_mut()`. Due to the `if x < width && y < height`, the compiler should notice that the bounds are already being checked, and elide/optimize out the second check. Even if the second check isn't optimized out, the performance impact should be negligible.

SorteKanin commented:
The compiler does not optimize out the second check - see here: https://godbolt.org/z/qbfPx7vP4

I understand the caution around unsafe code, but this is a very simple usage of unsafe. I can hardly think of a simpler example.

If you really don't want any usage of unsafe, consider adding ``#![forbid(unsafe_code)]` to `lib.rs`. That would signify that choice to contributors more clearly. But I would encourage you to perhaps consider if this case is not safe enough.

Razaekel commented:
You're absolutely right that if I want people to not use unsafe, I should have the forbid flag active. I'll have to add that in.

There are easier ways to get better performance without having to use the small optimization of using `get_unchecked()`, such as using rayon for parallel processing, or using SIMD and vectorizing the code. So, for now, please use `get()` instead of `get_unchecked()`, and the performance can be re-investigated once the easier stuff has been implemented.

SorteKanin commented:
I've removed the unsafe now :)

SorteKanin commented:
Also I realized that the `data_mut` function was a bad idea since the user could mutate the underlying vector to a different length in that case, breaking the invariants of the struct. I've changed it to implement iteration over the underlying vector instead, including mutable iteration, but not the ability to mutate the vector itself.

Razaekel commented:
I think that instead of panicking when attempting to set a value at a location that's out of bounds, it would be better to either silently discard the value, or print an error. After all, having IndexMut return the border value would give the caller the ability to change that value, which is not something I want to have happen except through set_border_value().

Alternatively, the border value could be cloned and returned, so that the caller does get the border value, but any changes to that doesn't affect the actual value stored in the struct.

SorteKanin commented:
> it would be better to either silently discard the value, or print an error.

I disagree. I feel such an error should definitely not be silently ignored (I consider printing an error basically silently ignoring it; you may not even discover the error at run-time). I feel that the idiomatic Rust way of doing things is to panic in that situation. This is what the standard `Vec` does after all. I also think that this is the expectation of the `Index` and `IndexMut` traits - they panic if you index outside the bounds (the border value is a special case of this specific data structure that allows it to not panic when indexed immutably).

You could provide a function similar to the `get` function on `Vec`s which returns an option instead of a reference that consumers of this library could then use if they need to index by a value that they are not sure are within the bounds.

Razaekel commented:
You have a good point, in regards to what Rust's `Vec` does.

In that case, this looks pretty good to me.
